[CONTRIBUTING.md] Add section about merging ECS conventions (#333)
open-telemetry/semantic-conventions: CONTRIBUTING.md

  - Send staging tag as PR for review.
- Create a tag `v{version}` on the merged PR and push remote.

## Merging existing ECS conventions

The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
Conventions per [OTEP 222][otep222]. When adding a semantic convention that
exists in some form in ECS, consider the following guidelines:

> - Prefer using the existing ECS name when possible. In particular:

Reviewer: Recently it was introduced container attributes from [OCI](https://github.com/open-telemetry/semantic-conventions/pull/159). What will be our guidelines when:

- Another standard (such OCI in this case) has attributes different from the ones from ECS? Will we still take ECS? Will ECS adapt to the other, potentially OSS standard?

Reviewer 2: I think that in https://github.com/open-telemetry/semantic-conventions/pull/159 we tried to follow the OCI spec and not directly use any naming convention. In this there was no breaking-change trade off, right?

Reviewer: Yes, I just mentioned that as an example of another open standard that could clash with ECS.

Reviewer 3: @joaopgrassi I think that would be rather an edge case that we would discuss case by case anyways, WDYT?

Author: I would take this as 'comparable evidence' on the rule for 'proposing a name that differs from the ECS convention'. Maybe we can explicitly add "prior work on a different standard" as a piece of evidence that can be 'used as evidence'?

Reviewer 4: In my understanding this case is well covered by this point:

> * If proposing a name that differs from the ECS convention, provide usage data, user issue reports or feature requests or comparable evidence about the alternatives.

If there's a "competing" standard, I think it needs to be evaluated on a case-by-case basis whether the additional benefits for users of being consistent with the "competing" standard outweigh the pain of breaking ECS.

Author: Added this explicitly on that bullet point for clarity

  - Do not amend ECS names to follow plural guidelines or other soft 

>     recommendations.

Reviewer 5: do we know how many of these there will be? if there's a lot, maybe we should just drop the pluralization guideline completely (#267)?

(not that i'm in favor of dropping the guideline, but if there's a lot of these then it feels like we will be essentially defacto dropping it)

Author: I think we can handle the discussion on dropping the specific rule separately. The main motivation for this line is the discussion in #203, but the rule is also meant to apply to the [counter naming rules](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#naming-rules-for-counters-and-updowncounters), the rules listed [here](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#instrument-naming) and other rules that are not a 'MUST'. In general, the rule I am trying to write down here is "it is okay to ignore a 'SHOULD' rule when integrating existing ECS conventions for the sake of backwards compatibility with ECS".  I think these rules are still useful to have to guide contributions that do not have an equivalent in ECS.

Reviewer 3: > not that i'm in favor of dropping the guideline, but if there's a lot of these then it feels like we will be essentially defacto dropping it

I think having that guideline as a `soft` recommendation still makes sense for any new attributes and where there are no strong arguments against it.

Author: @trask This has been reworded in the last iteration, PTAL

Author: I am marking this as solved since we rediscussed this on a discussion below.

>     attribute naming rules.

Reviewer 6: It's a bit unclear what are the "plural guidelines or other syntactical attribute naming rules".

Also, is this file the best place to define it? Maybe it should go to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md where all the guidelines are defined? It seems like we are scattering the guidelines across different places

Reviewer 2: The linked file is metrics specific though, no?
I had found https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md#name-pluralization-guidelines about the pluralization rule which seems to be more generic?

Maybe https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/README.md is a more generic one for this repo. I'm not sure if it's searchable in the docs page though.

For now I would be fine with keeping it in the contribution guide as well.

Reviewer 3: @dmitryax does that resolve your comment?

Author: @dmitryax can we resolve this?

  - Send staging tag as PR for review.
- Create a tag `v{version}` on the merged PR and push remote.

## Merging existing ECS conventions

The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
Conventions per [OTEP 222][otep222]. When adding a semantic convention that
exists in some form in ECS, consider the following guidelines:

- Prefer using the existing ECS name when possible. In particular:

>   - Do not amend ECS names to follow plural guidelines or other soft

Reviewer: I guess the goal of this sentence is to say something like: "Don't change the ECS names just to comply with OTel attribute guidelines" right? Like the pluralization rules.

In that case, what about we go with something more generic:

```suggestion
  - ECS names should not be modified to comply with OpenTelemetry general attribute guidelines. 
```

OR
```suggestion  
  - Special consideration must be applied for ECS attribute names that do not comply with the OpenTelemetry general attribute guidelines. 
```

OR

```suggestion
 - If an ECS attribute name does not comply with the OpenTelemetry general attribute guidelines and changing it is not desirable, provide data to support why changing the ECS attribute change cannot be done.
```

These are just suggestions.

Reviewer 3: @joaopgrassi I like that direction!

How about?:

> ECS names should not be modified for the _only reason_ of complying with OpenTelemetry general attribute guidelines.

  - Send staging tag as PR for review.
- Create a tag `v{version}` on the merged PR and push remote.

## Merging existing ECS conventions

The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
Conventions per [OTEP 222][otep222]. When adding a semantic convention that
exists in some form in ECS, consider the following guidelines:

- Prefer using the existing ECS name when possible. In particular:
  - Do not modify an ECS name for the only reason of complying with

>     OpenTelemetry general attribute guidelines.

Reviewer 7: Based on the discussions on `host.ip` we might want to define this a bit more.

Author: Do you mean refining as in "we should make it clear what the general attribute guidelines are" or "specifying which attribute guidelines we should follow"?

Reviewer: with "general attribute guidelines" I was mostly referring to this: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines. That is the central place we have such guidelines.. and if something is not there, we should add. (for ex #394 is not there yet)

Author: The recommendations here https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md should also be part of this IMO. If you think it makes sense to explicitly list what guidelines we are talking about I can add an explicit list.

Reviewer 3: @joaopgrassi I think the guidelines @mx-psi linked (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md) are more generic and (should) reference the one you linked.

https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines is only about metrics, right?

Reviewer: Right, all of those apply, yes. I think it would be good to link them in here.

>     OpenTelemetry general attribute guidelines.

Reviewer 3: ```suggestion
    [OpenTelemetry general attribute naming guidelines](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/).
```

>     [OpenTelemetry general attribute naming guidelines](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/).

Reviewer: We discussed this PR in the working group meeting on Nov 27. The general feeling is that this bullet point is "too permissive" and pretty much opens up to disregard completely the guidelines in place, including the ones which are important (e.g. character set).

We discussed these options/alternatives/improvements

1. Adapt this bullet point to better indicate which guidelines shouldn't be applied, or adapt the text such as it is not so broad 
2. Remove this paragraph completely
3. Close this PR and deal with problematic/conflicting ECS attributes as they appear
4. Indicate here that new attributes should be checked on ECS first if they exist there before suggesting new attributes in OTel to avoid even more conflicts

The conclusion was we should aim to do a mix of 1, and 4

Reviewer 8: Great summary @joaopgrassi 

I think logic-wise it's better start with second bullet point 
```
-  If proposing a name that differs from the ECS convention, provide usage
    data, user issue reports, feature requests, examples of prior work on a
    different standard or comparable evidence about the alternatives.
```
And then add more clarification, because most of the problems coming from `Name Pluralization guidelines`, everything else from naming guidelines are the same for ECS as well
```
- When no suitable alternatives are provided, altering an ECS name solely 
  for the purpose of complying with [Name Pluralization guidelines] (https://opentelemetry.io/docs/specs/otel/common/attribute-naming/#name-pluralization-guidelines) 
  SHOULD BE avoided.
- Before proposing a new attribute, it's advisable to check ECS first to ensure 
  a similar attribute doesn't already exist. If a match is found, consider adopting it 
  into Otel to minimize potential conflicts.
```

Reviewer 5: i'd suggest slightly more permissive still: `SHOULD BE avoided` -> `MAY BE avoided`

Author: Resolving per approval from Joao :)

  - Send staging tag as PR for review.
- Create a tag `v{version}` on the merged PR and push remote.

## Merging existing ECS conventions

The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
Conventions per [OTEP 222][otep222]. When adding a semantic convention that
exists in some form in ECS, consider the following guidelines:

- Prefer using the existing ECS name when possible. In particular:
  - If proposing a name that differs from the ECS convention, provide usage
    data, user issue reports, feature requests, examples of prior work on a
    different standard or comparable evidence about the alternatives.
  - When no suitable alternatives are provided, altering an ECS name solely

>     for the purpose of complying with [Name Pluralization guidelines](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/#name-pluralization-guidelines)

Reviewer 9: ```suggestion
    for the purpose of complying with [Name Pluralization guidelines](docs/general/attribute-naming.md#name-pluralization-guidelines)
```
These guidelines have moved into this repo here.

    MAY BE avoided.
- Do not use an existing ECS name as a namespace. If the name must differ, use a
  different namespace name to avoid clashes or avoid using the namespace

>   entirely.

Reviewer 9: ```suggestion
  entirely. See the [ECS field reference] for existing namespaces.
```

[nvm]: https://github.com/nvm-sh/nvm/blob/master/README.md#installing-and-updating
[stability guarantees]: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.26.0/specification/versioning-and-stability.md#semantic-conventions-stability

> [otep222]: https://github.com/open-telemetry/oteps/pull/222

Reviewer 9: ```suggestion
[otep222]: https://github.com/open-telemetry/oteps/pull/222
[ECS field reference]: https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html
```
